src/Hierarchy.elm

```diff
@@ -0,0 +1,267 @@
+    else
+        let
+            -- Finds out the number of nodes that we've already registered
+            -- that claim to be our child.
+            currentCount =
+                h.childIndex
+                    |> Dict.get folder.id
+                    |> Maybe.map List.length
+                    |> Maybe.withDefault 0
+
+            nodeDBWithFolder =
+                Dict.insert folder.id (Folder currentCount folder) h.nodeDB
+
+            -- XXX If our current children count is 0, we add ourselves
+            -- to the child index because it's possible no node will get
+            -- registered as our child.
+            childIndexWithFolder =
+                if currentCount == 0 then
+                    Dict.insert folder.id [] h.childIndex
+                else
+                    h.childIndex
+
+            ( newRootNodes, newChildIndex, newNodeDB ) =
+                case folder.parent of
+                    Nothing ->
+                        ( folder.id :: h.rootNodes, childIndexWithFolder, nodeDBWithFolder )
+
+                    Just p ->
+                        ( h.rootNodes
+                        , Dict.update p (addChild folder.id) childIndexWithFolder
+                        , incrementCounts p nodeDBWithFolder
+                        )
+        in
+            Ok <|
+                Hierarchy
+                    { h
+                        | nodeDB = newNodeDB
+                        , childIndex = newChildIndex
+                        , rootNodes = newRootNodes
+                    }
+
+
+{-| Add a file to the hierarchy
+-}
+registerFile : Audience -> Hierarchy -> Result String Hierarchy
+registerFile file (Hierarchy h) =
+    if Dict.member file.id h.nodeDB then
+        Err <| "Node with ID " ++ (toString file.id) ++ " has already been added"
+    else
+        let
+            nodeDBWithFile =
+                Dict.insert file.id (File file) h.nodeDB
+
+            ( newRootNodes, newChildIndex, newNodeDB ) =
+                case file.folder of
+                    Nothing ->
+                        ( file.id :: h.rootNodes, h.childIndex, nodeDBWithFile )
+
+                    Just p ->
+                        ( h.rootNodes
+                        , Dict.update p (addChild file.id) h.childIndex
+                        , incrementCounts p nodeDBWithFile
+                        )
+        in
+            Ok <|
+                Hierarchy
+                    { h
+                        | nodeDB = newNodeDB
+                        , childIndex = newChildIndex
+                        , rootNodes = newRootNodes
+                    }
+
+
+{-| Adds a node ID to the list of node references
+This is a helper function used in a `Dict.update` call applied to the
+children index.
+-}
+addChild : Int -> Maybe (List Int) -> Maybe (List Int)
+addChild id maybeList =
+    maybeList
+        |> Maybe.map ((::) id)
+        |> Maybe.orElseLazy (\_ -> Just [ id ])
```

Comment:
There it will be much simplier to have `List a` instead of `Maybe (List a)` since `Nothing` and `Just []` are kind of similar in this case. Generally I think using `Maybe (List a)` makes sense only in applicative monadic or similar operations.